crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -426,15 +426,106 @@ impl Format<PyFormatContext<'_>> for MaybeParenthesizeExpression<'_> {
                     expression.format().with_options(Parentheses::Never).fmt(f)
                 }
                 Parenthesize::IfBreaks => {
-                    if node_comments.has_trailing() {
-                        expression.format().with_options(Parentheses::Always).fmt(f)
+                    // Is the expression the last token in the parent statement.
+                    // Excludes `await` and `yield` for which Black doesn't seem to apply the layout?
+                    let last_expression = parent.is_stmt_assign()
+                        || parent.is_stmt_ann_assign()
+                        || parent.is_stmt_aug_assign()
+                        || parent.is_stmt_return();
+
+                    // Format the statements and value's trailing end of line comments:
+                    // * after the expression if the expression needs no parentheses (necessary or the `expand_parent` makes the group never fit).
+                    // * inside the parentheses if the expression exceeds the line-width.
+                    //
+                    // ```python
+                    // a = long # with_comment
+                    // b = (
+                    //     short # with_comment
+                    // )
+                    //
+                    // # formatted
+                    // a = (
+                    //     long # with comment
+                    // )
+                    // b = short # with comment
+                    // ```
+                    // This matches Black's formatting with the exception that ruff applies this style also for
+                    // attribute chains and non-fluent call expressions. See https://github.com/psf/black/issues/4001#issuecomment-1786681792
+                    //
+                    // This logic isn't implemented in [`place_comment`] by associating trailing statement comments to the expression because
+                    // doing so breaks the suite empty lines formatting that relies on trailing comments to be stored on the statement.
```

Comment:
Would this work out-of-the-box if we _did_ associate the trailing statement comments on the expression, ignoring the suite formatting?

Reply:
No, it would still require the `if_group_breaks` and `if_group_fits` but it would avoid the unprecedented "the child steals comments from its parent".